File: Emulator/src/main/core/game/node/entity/lock/Lock.java
package core.game.node.entity.lock;

import core.game.world.GameWorld;

/**
 * The type Lock.
 */
public class Lock {

    private int expiration;

    private LockElapse elapse;

    private String message;

    /**
     * Instantiates a new Lock.
     */
    public Lock() {
        this(null);
    }

    /**
     * Instantiates a new Lock.
     *
     * @param message the message
     */
    public Lock(String message) {
        this.message = message;
    }

    /**
     * Lock.
     */
    public void lock() {
        lock(Integer.MAX_VALUE - GameWorld.getTicks());
    }

    /**
     * Lock.
     *
     * @param ticks the ticks
     */
    public void lock(int ticks) {
        if (ticks > expiration - GameWorld.getTicks()) {
            this.expiration = GameWorld.getTicks() + ticks;
        }
    }

    /**
     * Unlock.
     */
    public void unlock() {
        this.expiration = 0;
    }

    /**
     * Is locked boolean.
     *
     * @return the boolean
     */
    public boolean isLocked() {
        return expiration > GameWorld.getTicks();
    }

    /**
     * Sets elapse.
     *
     * @param elapse the elapse
     * @return the elapse
     */
    public Lock setElapse(LockElapse elapse) {
        this.elapse = elapse;
        return this;
    }

    /**
     * Gets elapse event.
     *
     * @return the elapse event
     */
    public LockElapse getElapseEvent() {
        return elapse;
    }

    /**
     * Gets message.
     *
     * @return the message
     */
    public String getMessage() {
        return message;
    }

    /**
     * Sets message.
     *
     * @param message the message
     */
    public void setMessage(String message) {
        this.message = message;
    }

}

File: server/src/main/core/game/node/entity/lock/Lock.kt
package core.game.node.entity.lock

/**
 * Represents a time-based lock used to restrict actions for a specified duration (in game ticks).
 *
 * @property message (Optional) message shown when the lock is active.
 */
class Lock(var message: String? = null) {

    private var expiration: Int = 0
    private var elapse: LockElapse? = null

    /**
     * Locks this instance for the specified number of ticks.
     *
     * @param ticks Number of game ticks to lock.
     */
    fun lock(ticks: Int) {
        val currentTicks = core.game.world.GameWorld.ticks
        if (ticks > expiration - currentTicks) {
            expiration = currentTicks + ticks
        }
    }

    /**
     * Unlocks this instance immediately.
     */
    fun unlock() {
        expiration = 0
    }

    /**
     * Returns whether this instance is currently locked.
     *
     * @return `true` if locked, otherwise `false`.
     */
    fun isLocked(): Boolean = expiration > core.game.world.GameWorld.ticks

    /**
     * Sets the elapse event for this lock.
     *
     * @param elapse The elapse callback.
     * @return This instance for chaining.
     */
    fun setElapse(elapse: LockElapse?): Lock {
        this.elapse = elapse
        return this
    }

    /**
     * Gets the elapse event associated with this lock.
     *
     * @return The [LockElapse] instance or `null`.
     */
    fun getElapseEvent(): LockElapse? = elapse
}
